test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,511 @@
+
+        class SGDWithNewKey(torch.optim.SGD):
+            # Dummy optimizer which adds a new key to the param groups
+            def step(self, closure=None):
+                super().step()
+                self.param_groups[0]["new_key"] = 0.1
+
+        x = torch.tensor([1.0], device=DEVICE, requires_grad=True)
+        o = ZeroRedundancyOptimizer([x], optim=SGDWithNewKey, lr=0.1)
+        x.backward()
+        o.step()
+        self.assertEqual(o.param_groups[0]["new_key"], 0.1)
+        self.assertEqual(x, torch.tensor([0.9], device=DEVICE))
+
+    def test_step_without_closure(self):
+        self.dist_init(self.rank)
+
+        class SGDWithoutClosure(torch.optim.SGD):
+            def step(self):
+                return super().step()
+
+        x = torch.tensor([1.0], device=DEVICE, requires_grad=True)
+        o = ZeroRedundancyOptimizer([x], optim=SGDWithoutClosure, lr=0.1)
+        x.backward()
+        o.step()
+        self.assertEqual(x, torch.tensor([0.9], device=DEVICE))
+
+    def test_local_state_dict(self):
+        self.dist_init(self.rank)
+
+        x = torch.tensor([1.0], device=DEVICE, requires_grad=True)
+        o = ZeroRedundancyOptimizer([x], optim=SGD, lr=0.1)
+        local_state_dict = o.local_state_dict()
+        o = ZeroRedundancyOptimizer([x], optim=SGD, lr=0.01)
+        o.load_local_state_dict(local_state_dict)
+        # We should now be using a lr of 0.1.
+        self.assertEqual(o.optim.param_groups[0]["lr"], 0.1)
+        self.assertEqual(o.param_groups[0]["lr"], 0.1)
+        x.backward()
+        o.step()
+        self.assertEqual(x, torch.tensor([0.9], device=DEVICE))
+
+    def test_implicit_local_state_dict(self):
+        self.dist_init(self.rank)
+
+        x = torch.tensor([1.0], device=DEVICE, requires_grad=True)
+        o = ZeroRedundancyOptimizer([x], optim=SGD, lr=0.1)
+        local_state_dict = o.state_dict()
+        o = ZeroRedundancyOptimizer([x], optim=SGD, lr=0.01)
+        o.load_state_dict(local_state_dict)
+        # We should now be using a lr of 0.1.
+        self.assertEqual(o.optim.param_groups[0]["lr"], 0.1)
+        self.assertEqual(o.param_groups[0]["lr"], 0.1)
+        x.backward()
+        o.step()
+        self.assertEqual(x, torch.tensor([0.9], device=DEVICE))
+
+
+class TestZeroRedundancyOptimizerTwoRanks(TestZeroRedundancyOptimizer):
+    @property
+    def world_size(self):
+        return 2
+
+    def test_zero_grad(self):
+        self.dist_init(self.rank)
+        x = torch.rand(1)
+        m = torch.nn.Linear(1, 1)
+        o = ZeroRedundancyOptimizer(m.parameters(), optim=SGD, lr=0.1)
+        y = m(x)
+        y.backward(x)
+        self.assertNotEqual(m.weight.grad, torch.zeros_like(m.weight))
+        self.assertNotEqual(m.weight.grad, torch.zeros_like(m.weight))
+        o.zero_grad()
+        self.assertFalse(m.weight.grad)
+        self.assertFalse(m.bias.grad)
+
+
+class TestZeroRedundancyOptimizerThreeRanks(TestZeroRedundancyOptimizer):
```

Comment:
Not sure we why have such test cases based on ranks, can't you have a fixed world_size (say 2 or 4) and run all tests on that?

Reply:
legacy, fixing that, you're right I think